sdk/core/azure-core/inc/azure/core/http/websockets/curl_websockets_transport.hpp

```diff
@@ -0,0 +1,144 @@
+        : CurlTransport(options)
+    {
+    }
+    /**
+     * @brief Implements interface to send an HTTP Request and produce an HTTP RawResponse
+     *
+     * @param request an HTTP Request to be send.
+     * @param context A context to control the request lifetime.
+     *
+     * @return unique ptr to an HTTP RawResponse.
+     */
+    virtual std::unique_ptr<RawResponse> Send(Request& request, Context const& context) override;
+
+    /**
+     * @brief Indicates if the transports supports native websockets or not.
+     *
+     * @details For the CURL websocket transport, the transport does NOT support native websockets -
+     * it is the responsibility of the client of the WebSocketTransport to format WebSocket protocol
+     * elements.
+     */
+    virtual bool NativeWebsocketSupport() override { return false; }
+
+    virtual void CompleteUpgrade() override;
+
+    /**
+     * @brief Gracefully closes the WebSocket, notifying the remote node of the close reason.
+     *
+     */
+    virtual void Close() override;
+
+    // Native WebSocket support methods.
+    /**
+     * @brief Gracefully closes the WebSocket, notifying the remote node of the close reason.
+     *
+     * @details Not implemented for CURL websockets because CURL does not support native websockets.
+     *
+     */
+    virtual void CloseSocket(uint16_t, std::string const&, Azure::Core::Context const&) override
+    {
+      throw std::runtime_error("Not implemented.");
+    }
+
+    /**
+     * @brief Retrieve the status of the close socket operation.
+     *
+     * @details Not implemented for CURL websockets because CURL does not support native websockets.
+     *
+     */
+    std::pair<uint16_t, std::string> GetCloseSocketInformation(const Azure::Core::Context&) override
+    {
+      throw std::runtime_error("Not implemented");
+    }
+
+    /**
+     * @brief Send a frame of data to the remote node.
+     *
+     * @details Not implemented for CURL websockets because CURL does not support native websockets.
+     *
+     */
+    virtual void SendFrame(
+        WebSocketFrameType,
+        std::vector<uint8_t> const&,
+        Azure::Core::Context const&) override
+    {
+      throw std::runtime_error("Not implemented.");
+    }
+
+    /**
+     * @brief Receive a frame of data from the remote node.
+     *
+     * @details Not implemented for CURL websockets because CURL does not support native websockets.
+     *
+     */
+    virtual std::vector<uint8_t> ReceiveFrame(WebSocketFrameType&, Azure::Core::Context const&)
+        override
+    {
+      throw std::runtime_error("Not implemented");
+    }
+
+    // Non-Native WebSocket support.
+    /**
+     * @brief This function is used when working with streams to pull more data from the wire.
+     * Function will try to keep pulling data from socket until the buffer is all written or until
+     * there is no more data to get from the socket.
```

Comment:
This is somehow different to the way BodyStream works.  Calling Read() from a stream would just pull any bytes available at the moment.

The issue with trying to read until buffer is all written or no more data to get is that we could block forever here.

Reply:
This isn't a BodyStream. It's a transport. And this is exactly the documentation of `CurlNetworkConnection::ReadFromSocket` - in fact this function is just a wrapper on `CurlNetworkConnection::ReadFromSocket`.